hardhat.config.ts

```diff
@@ -55,13 +110,80 @@ const config: HardhatUserConfig = {
       },
     },
   },
-  gasReporter: {
-    enabled: process.env.REPORT_GAS !== undefined,
-    currency: "USD",
+  baseURIs: {
+    // mainnets
+    ethereum: "https://tableland.network/tables/",
+    optimism: "https://tableland.network/tables/",
+    polygon: "https://tableland.network/tables/",
+    // testnets
+    "ethereum-rinkeby": "",
+    "ethereum-goerli": "https://testnet.tableland.network/tables/",
+    "optimism-kovan": "https://testnet.tableland.network/tables/",
+    "polygon-mumbai": "https://testnet.tableland.network/tables/",
+    // devnets
+    "ethereum-rinkeby-staging": "",
+    "optimism-kovan-staging": "https://staging.tableland.network/tables/",
+    localhost: "http://localhost:8080/tables/",
   },
-  etherscan: {
-    apiKey: process.env.ETHERSCAN_API_KEY || "",
+  proxies: {
+    // tableland mainnet mainnets
+    ethereum: "",
+    optimism: "",
+    polygon: "",
+    // tableland testnet testnets
+    "ethereum-rinkeby": "0x30867AD98A520287CCc28Cde70fCF63E3Cdb9c3C", // deprecating: do not upgrade!
+    "ethereum-goerli": "",
+    "optimism-kovan": "",
+    "polygon-mumbai": "",
+    // tableland staging testnets
+    "ethereum-rinkeby-staging": "0x847645b7dAA32eFda757d3c10f1c82BFbB7b41D0", // deprecating: do not upgrade!
+    "optimism-kovan-staging": "",
+    localhost: "",
   },
-};
+}
+
+interface TablelandNetworkConfig {
+  // mainnets
+  ethereum: string
+  optimism: string
+  polygon: string
+
+  // testnets
+  "ethereum-rinkeby": string // deprecating
+  "ethereum-goerli": string
+  "optimism-kovan": string
+  "polygon-mumbai": string
+
+  // devnets
+  "ethereum-rinkeby-staging": string // deprecating
+  "optimism-kovan-staging": string
+  localhost: string // hardhat
+}
+
+declare module "hardhat/types/config" {
+  // eslint-disable-next-line no-unused-vars
+  interface HardhatUserConfig {
+    baseURIs: TablelandNetworkConfig
+    proxies: TablelandNetworkConfig
+  }
+}
+
+declare module "hardhat/types/runtime" {
+  // eslint-disable-next-line no-unused-vars
+  interface HardhatRuntimeEnvironment {
+    baseURI: string
+    proxy: string
+  }
+}
+
+extendEnvironment((hre: HardhatRuntimeEnvironment) => {
```

Comment:
extend the `hre` so that you can simply run `scripts/deploy.ts` or `scripts/upgrade.ts` and have the correct base URI and proxy injected.

Reply:
cool